src/middleware/from-channel.js

```diff
@@ -0,0 +1,23 @@
+export default context => {
+  if (!process.client) return;
+
+  const {
+    from: fromRoute,
+    route: toRoute,
+    redirect,
+    getRouteBaseName,
+  } = context;
+
+  // If the from query is present in the current route but not in the next route, add it to the next route
+  if (
+    fromRoute.query.from &&
+    !toRoute.query.from &&
+    // Exclude navigating from /nft/* routes
+    !getRouteBaseName(fromRoute).startsWith('nft')
```

Comment:
If we set applicable condition here, I think this middleware can continue to be global